solidity/contracts/bridge/Bridge.sol

```diff
@@ -794,10 +955,577 @@ contract Bridge is Ownable {
+    //         HASH160 over the compressed ECDSA public key) of the wallet which
+    ///        performed the redemption transaction.
+    /// @return info Outcomes of the processing.
+    function processRedemptionTxOutputs(
+        bytes memory redemptionTxOutputVector,
+        bytes20 walletPubKeyHash
+    ) internal returns (RedemptionTxOutputsInfo memory info) {
+        // Determining the total number of redemption transaction outputs in
+        // the same way as for number of inputs. See `BitcoinTx.outputVector`
+        // docs for more details.
+        (uint256 outputsCompactSizeUintLength, uint256 outputsCount) =
+            redemptionTxOutputVector.parseVarInt();
+
+        // To determine the first output starting index, we must jump over
+        // the compactSize uint which prepends the output vector. One byte
+        // must be added because `BtcUtils.parseVarInt` does not include
+        // compactSize uint tag in the returned length.
+        //
+        // For >= 0 && <= 252, `BTCUtils.determineVarIntDataLengthAt`
+        // returns `0`, so we jump over one byte of compactSize uint.
+        //
+        // For >= 253 && <= 0xffff there is `0xfd` tag,
+        // `BTCUtils.determineVarIntDataLengthAt` returns `2` (no
+        // tag byte included) so we need to jump over 1+2 bytes of
+        // compactSize uint.
+        //
+        // Please refer `BTCUtils` library and compactSize uint
+        // docs in `BitcoinTx` library for more details.
+        uint256 outputStartingIndex = 1 + outputsCompactSizeUintLength;
+
+        // Calculate the keccak256 from `walletPubKeyHash` upfront to avoid
+        // computing it multiple times within the processing loop and save
+        // on gas costs in result.
+        bytes32 walletPubKeyHashKeccak =
+            keccak256(abi.encodePacked(walletPubKeyHash));
+
+        // Outputs processing loop.
+        for (uint256 i = 0; i < outputsCount; i++) {
+            // Check if we are at the end of the output vector.
+            if (outputStartingIndex >= redemptionTxOutputVector.length) {
+                break;
+            }
+
+            // TODO: Check if we can optimize gas costs by adding
+            //       `extractValueAt` and `extractHashAt` in `bitcoin-spv-sol`
+            //       in order to avoid allocating bytes in memory.
+            uint256 outputLength =
+                redemptionTxOutputVector.determineOutputLengthAt(
+                    outputStartingIndex
+                );
+            bytes memory output =
+                redemptionTxOutputVector.slice(
+                    outputStartingIndex,
+                    outputLength
+                );
+
+            // Extract the value and hash from given output.
+            uint64 outputValue = output.extractValue();
+            bytes memory outputHash = output.extractHash();
+
+            if (
+                info.changeValue == 0 &&
```

Comment:
So to optimize the gas cost of redemption proof, we could expect that the change is always the first output, right? 

If so, I think it's worth adding to `@dev` section of `submitRedemptionProof` function.

Reply:
Yep indeed. See 1ae7b1d3